README.adoc

```diff
@@ -1,180 +1,192 @@
 
-----
-$ bundle update
-----
+By letting Locust operate with 2000 users for a few minutes, I get a RSS of about *96 MB*.
 
-The dependencies are specified in the `Gemfile` file:
+With OpenJDK 11 and no tuning, I get a RSS of about *247 MB*.
+
+Let us use OpenJDK 11 again but now with some limits:
 
-[source,ruby]
 ----
-include::Gemfile[]
+$ java -Xmx32m -Xss256k -jar build/libs/openj9-howto-all.jar
 ----
 
-TIP: Keep the `vertx-howtos-jekyll-theme` dependency up-to-date!
-
-=== Running Jekyll
+RSS usage goes to *124 MB* on my machine.
 
-You can run Jekyll with a full auto-reload experience:
+Now let us go back to OpenJ9 with a similar configuration:
 
 ----
-$ bundle exec jekyll server --watch --livereload
+$ java -Xmx32m -Xss256k -Xshareclasses -Xshareclasses:name=sum -Xshareclasses:cacheDir=_cache -jar build/libs/openj9-howto-all.jar
 ----
 
-== How is the website being published?
+RSS usage is about *73 MB* on my machine.
 
-The content first needs to be rendered by Jekyll, then the resulting web site files are being pushed to the repository `gh-pages` branch.
-GitHub pages then exposes the content of the `gh-pages` branch as the website.
+== Building and running a Docker image
 
-While you could do this manually, we automate the process with Travis-CI that also has a special deployment provider to further simplify the process.
+Ok so we have seen how gentle OpenJ9 was on memory even without tuning.
+Let us now package the micro-service as a Docker image.
 
-=== Travis-CI configuration
+Here is the `Dockerfile` you can use:
 
-Feel-free to customize the `.travis.yml` configuration to check that your how-to code builds and passes tests, but keep the `deploy` tree close to that one:
-
-[source,yaml]
+[source,text]
 ----
-include::.travis.yml[]
+include::Dockerfile[]
 ----
 
-`$GITHUB_TOKEN` is a Travis-CI secret to pass a GitHub API token so that the CI job can push to the `gh-pages` branch.
-It is configured on Travis-CI by a Vert.x project maintainer.
-
-Note that we only publish on the `master` branch, but pull-requests can still be tested.
-This also means that we are in a model where the `master` branch is the latest stable revision.
-
-=== Travis-CI cron job
+You can note:
 
-It is strongly advised to set-up a cron job in Travis-CI to regularly build a how-to.
+- `-Xvirtualized` is a flag for virtualized / container environments so OpenJ9 reduces CPU consumption when idle
+- `/app/_cache` is a volume that will have to be mounted for containers to share the OpenJ9 classes cache.
 
-Building on a daily basis is a good option to ensure theme updates are quickly reflected.
+The image can be built as in:
 
-== Misc
+----
+$ docker build . -t openj9-app
+----
 
-=== Code style
+We can then create containers from the image:
 
-There is a `.editorconfig` file in this template repository with the 2-space indentation style that we prefer in Vert.x.
+----
+$ docker run -it --rm -v /tmp/_cache:/app/_cache -p 8080:8080 openj9-app
```

Comment:
On some platforms with `cgroups` active the container cannot access the shared classes directory. They will get an error like:

```
JVMSHRC226E Error opening shared class cache file 
JVMSHRC336E Port layer error code = -102
JVMSHRC337E Platform error message: Permission denied
JVMJ9VM015W Initialization error for library j9shr29(11): JVMJ9VM009E J9VMDllMain failed
Error: Could not create the Java Virtual Machine.
Error: A fatal exception has occurred. Program will exit.
```

Can you add a `TIP` about this and say they can relax the check with `-v /tmp/_cache:/app/_cache:z`

Reply:
👍